packages/dev/core/src/Materials/Node/Blocks/mathBlock.ts

```diff
@@ -0,0 +1,108 @@
+    }
+
+    /**
+     * Gets the output component
+     */
+    public get output(): NodeMaterialConnectionPoint {
+        return this._outputs[0];
+    }
+
+    private _updateInputOutputTypes() {
+        // First update the output type with the initial assumption that we'll base it on the left input.
+        this.output._typeConnectionSource = this.left;
+
+        if (this.left.isConnected && this.right.isConnected) {
+            // Both inputs are connected, so we need to determine the output type based on the input types.
+            if (
+                this.left.type === NodeMaterialBlockConnectionPointTypes.Int ||
+                (this.left.type === NodeMaterialBlockConnectionPointTypes.Float && this.right.type !== NodeMaterialBlockConnectionPointTypes.Int)
+            ) {
+                this.output._typeConnectionSource = this.right;
+            }
+        } else if (this.left.isConnected !== this.right.isConnected) {
+            // Only one input is connected, so we need to determine the output type based on the connected input.
+            this.output._typeConnectionSource = this.left.isConnected ? this.left : this.right;
+        }
+
+        // Next update the accepted connection point types for the inputs based on the current input connection state.
+        if (this.left.isConnected || this.right.isConnected) {
+            for (const [first, second] of [
+                [this.left, this.right],
+                [this.right, this.left],
+            ]) {
+                // Always allow Ints and Floats.
+                first.acceptedConnectionPointTypes = [NodeMaterialBlockConnectionPointTypes.Int, NodeMaterialBlockConnectionPointTypes.Float];
+
+                if (second.isConnected) {
+                    // The same types as the connected input are always allowed.
+                    first.acceptedConnectionPointTypes.push(second.type);
+
+                    // If the other input is a scalar, then we also allow Vector/Color/Matrix types.
+                    if (second.type === NodeMaterialBlockConnectionPointTypes.Int || second.type === NodeMaterialBlockConnectionPointTypes.Float) {
+                        first.acceptedConnectionPointTypes.push(
+                            NodeMaterialBlockConnectionPointTypes.Vector2,
+                            NodeMaterialBlockConnectionPointTypes.Vector3,
+                            NodeMaterialBlockConnectionPointTypes.Vector4,
+                            NodeMaterialBlockConnectionPointTypes.Color3,
+                            NodeMaterialBlockConnectionPointTypes.Color4,
+                            NodeMaterialBlockConnectionPointTypes.Matrix
+                        );
+                    }
+                }
+            }
+        }
+    }
+
+    /**
+     * Release resources
+     */
+    public override dispose(): void {
+        super.dispose();
+        this._connectionObservers.forEach((observer) => observer.remove());

```

Comment:
I guess you also need to clear the this._connectionObservers array

Reply:
`_connectionObservers` is intended to essentially be a tuple of 4 observers, and to be the equivalent of having 4 properties on the class. It's just that storing them as an array to slightly simplify the logic. Clearing the "array" would be equivalent to setting 4 separate properties to null/undefined, which seems a like maybe dispose overkill. I could do it, but to do so I need to make the code more permissive (e.g. mutable array). Personally I think being more restrictive is more valuable because more restrictive code has fewer opportunities for bugs, but I don't feel super strongly about it. If you feel more strongly about clearing the array, I can either make it not readonly, or split it out into 4 properties so there is no array to be cleared, just let me know!

Reply:
The problem is GC, If someone holds a link to a mathBlock, if you dispose the block there is still a link between the this._connectionObservers and the observer who in turn are keeping link to their "Action"

Or do I miss something?

Reply:
The observers are all state contained within the mathBlock itself (not external), so as it is that state/memory stays alive with the mathBlock (same as values of other properties of the class, for example). Clearing this collection would allow some of the memory that makes up the class be GC'd early if you still held a reference to the mathBlock. That could be an optimization, and in my mind that would be the same optimization as if we said we should always set all properties of a class (that reference any kind of object) to null upon dispose, which I don't think we do elsewhere. For example, nodeMaterialBlock does not set _buildTarget to null in dispose. Similarly, if instead of holding these observers in an array I held them in individual properties (e.g. leftConnectionObservable, leftDisconnectionObservable, rightConnectionObservable, rightDisconnectionObservable), I wouldn't expect the dispose function to set all those properties to null (the equivalent of clearing the collection).

Anyway, just providing that as explanation of my thinking. I'm not terribly opposed to changing it, it just means I need to remove the readonly-ness of the collection which obscures the intent of the code a bit.

Reply:
I disagree :)
Every observable are cleared on dispose because they can hold external references
https://github.com/BabylonJS/Babylon.js/blob/master/packages/dev/core/src/scene.ts#L4905

And you don't need to remove readOnly (i think) if you set array.length = 0

Reply:
Also, it might be more clear that I'm basically just mashing 4 properties into 1 if I changed the type to:
`private readonly _connectionObservers: Tuple<Observer<NodeMaterialConnectionPoint>, 4>;`